### ISSUE TITLE
Custom password adapter

### ISSUE THREAD

**Issue by TimNZ** [parse-community/parse-server#4406]
Before I do the dev and pull request, feedback please.

I want to support a custom password adapter in case I migrate to a different platform in the future and I don't want to hassle the users in having to do a password update if the new platform doesn't support hashed password import i.e. AWS Cognito
Putting aside any security concerns with what I might do, I am thinking:

* Add a 'passwordAdapter' to ParseServer config
- string = module to require
- object = instance

-> instance exports 'init', 'hash' and 'compare' methods 
init (if exists): init(parseServerConfig) { }
hash: as per password.js
compare: as per password.js

If no passwordAdapter specified, current password.js will be used.

lib/RestWrite.js and lib/Routers/UserRouter.js:
* use config.passwordAdapter instead of required 'password.js'

**Comment by montymxb** [comment]
This sounds like a pretty good start. I think it would be useful to factor out the current implementation into an adapter as well, making that the default. 

@flovilmart any on thoughts on something like this?

**Comment by flovilmart** [comment]
I’m not sure how it would solve your use case for cognito. Before we move anywhere with this kind of changes, I’d like to make sure that your intended way of using such adapter is secure,safe and doesn’t require additional maintenance.

**Comment by TimNZ** [comment]
Thanks for the concern.  

This discussion seems to encapsulate my motivations:
https://stackoverflow.com/questions/6980044/using-asymmetric-encryption-to-secure-passwords

I don't have any experience with Cognito, but a quick scan shows I can programmatically create new users in a pool by specifying username/password (cleartext) /email.

https://github.com/aws/amazon-cognito-identity-js

In this case since Parse Server is already modular with a lot of adapters, allowing override of opinionated defaults, shifting the current password mechanism to an adapter regardless is easy with any deliberate override at the discretion of the implementer.

I'm doing basic future proofing and as far as I can see only password storage is a potential problem for a future migration.

I want to launch with Parse now but if I get traction on my platform I expect to migrate and don't want pain for users.
This is why I'd only ever use payment platforms now that support PCI/secure transfers to another provider, because lock in sucks unless you decide you accept the lock in for whatever reason.

**Comment by flovilmart** [comment]
Parse server don’t store the password in clear, nor you should be able to reverse the password encryption. So I don’t see how you would move one user’s login from parse to cognito while not requiring them to replace it.

**Comment by TimNZ** [comment]
Sorry mate, confused about the position you've reached,
and my response is I hope constructive challenging of it, not an argument :)

I don't know know why you think I am planning to store passwords in clear text.

The link to the Stack Overflow discussion about asymmetric password storing and my summary were pretty clear, but you are saying things that make it seem you aren't reading what I wrote, or strangely interpreting it.  

As I said and linked to docs, you can create users in Cognito and specify a cleartext password using the API.

So the only issue is what's an acceptable compromise of securing passwords but being able to decrypt them if I need to.

What is the issue with moving the current password system to an adapter purely in context of Parse Server being customisable through adapters.  
I can do it and a push in 30 mins.

Should you be over reaching into how people choose to use it afterwards.
Is that your place - what's the right line?

The hashed passwords in Parse using bcrypt aren't uniquely salted as far as I can see, so that's not best practice either.  They can be bulk matched if ever compromised.

**Comment by flovilmart** [comment]
> I want to support a custom password adapter in case I migrate to a different platform in the future and I don't want to hassle the users in having to do a password update if the new platform doesn't support hashed password

I still don’t get how a custom password storage adapter will help you now, unless you start sending those clear text passwords to cognito now.

If that’s the case you want to implement, why not, I’m not against moving forward with the proposal. 
Also note that in terms of identity management, you can very well use the auth adapters and achieve a similar result. Ie: authenticate against cognito as a provider, and store a userId/token in the user table.

**Comment by flovilmart** [comment]
We’re not debating wether it’s technically feasible or not, but i’d like to get a better grasp of how you intent to use this feature. Wether it is by defereing from day 0 to a 3rd party or planing for a possible migration.

**Comment by TimNZ** [comment]
Ok, auth provider might be an alternative, but more work for Day 0.  I'll research further.

What I was envisaging was:

Day 0 - get up and running as quickly and easily as possible staying inside Parse: use out of box password support in Parse, but use a custom password adapter to support a different hashing method
Deferred: thinking about where I may wish to store user accounts and authentication info in the future

Future - decide to migrate to platform X, and load users into new platform X with passwords I am able to decrypt, which Cognito (as an example) supports.

If I use a custom password adapter that implements asymmetric encryption I only have to store the public key on the Parse Server for hashing/compare and the private key is somewhere else securely stored, and used to decrypt passwords in the future when needed.

**Comment by TimNZ** [comment]
The Stack Overflow link I included earlier explains the same motivations I have, this is why I'm confused as to what you are not getting what I'm trying to achieve.

https://stackoverflow.com/questions/6980044/using-asymmetric-encryption-to-secure-passwords

**Comment by flovilmart** [comment]
Be assured it’s not bad faith, no harm intended. Sometimes, things are not fully read, skipped, and i prefer having a proper explanation instead of a link to follow :)

**Comment by montymxb** [comment]
There tends to be a lot to go over for the servers and the sdks 😄 . Definitely helps out if you can walk us through things A -> Z as best you can.

**Comment by TimNZ** [comment]
Really appreciate the work and time you guys spend on this and the responsiveness. 
My comments don't reflect any sense of entitlement or ungratefulness, having said that..

I don't think I'm being unreasonable in saying if we'd kept the scope of the discussion to yes or no on moving password.js to PasswordAdapter so people could choose to implement their own hashing mechanism, which is a very contained change that I would have done the PR on, some time could have been saved :)

Don't think many devs want maintainers of software questioning too hard how they want to use it.

I know I've come in guns blazing wanting to change a bunch of things asap, so happy to be told to get in line and calm down.

**Comment by flovilmart** [comment]
We, as maintainers, have this duty to ask those questions. Otherwise we’d be accepting any change because it makes sense for one user. Any change, has to be thoroughly discussed as you may put it in today, but may let all hell loose after you decide you no longer need the project. While it’s true that this change look minimal from your perspective, you also have to understand ours. Touching something sensitive as this needs careful discussion before we can have a reflected decision. 
This process is normal, and it’s better to have this now than during your PR review.
